File: leetcode/algorithms/binary-search/search-insert-position.ts
/*
    leetcode
    category: "Algorithms"
    problem title: "Search Insert Position"
*/

function searchInsert(nums: number[], target: number): number {
    let high: number = nums.length-1;
    let low: number = 0;
    let index: number = -1
    while(high-low>1){
        let mid: number = Math.floor((high+low)/2);
        
        if(nums[mid]===target){
            index = mid;
            break;
        } else if(nums[mid]>target){
            high = mid;
        } else {
            low = mid;
        }
    }
    
    if(index===-1){
        if(nums[low]>=target){
            index=Math.max(0,low-1);
        } else if(nums[high]<target){
            index = high+1;
        } else {
            index = high;
        }
    }
    return index;
};

File: leetcode/algorithms/binary-search/search-insert-position.java
class Solution {
    public int searchInsert(int[] nums, int target) {
         int l = 0,h = nums.length - 1,tindex = -1;
        
        while(h-l>1){
            int m = l + ((h-l)/2);
            
            if(nums[m]==target){
                tindex = m;
                break;
            } else if(nums[m]<target){
                l = m;
            } else {
                h = m;
            }
        }
        
        if(tindex==-1){
            if(nums[l]>=target){
                tindex = l;
            } else if(nums[h]<target){
                tindex = h+1;
            } else {
                tindex = h;
            }
        }
        
        return tindex;
    }
} 